File: fe/core/domain/agent-providers.ts
// Whether each Planner provider can run on this server right now, and why not (#1817).

import { z } from 'zod';

import { agentProviderSchema, type AgentProvider } from '../api/schemas.js';
import type { ApiOperation } from '../api/types.js';

const checkedAtSchema = z.number();

/**
 * One provider's answer. `reason` is the server's own sentence with its fix, present exactly when the
 * provider is not `ready`; `not_configured` is a backend this server was not started with.
 */
export const providerAvailabilitySchema = z.discriminatedUnion('status', [
  z.object({ provider: agentProviderSchema, status: z.literal('ready'), reason: z.null(), checked_at_ms: checkedAtSchema }),
  z.object({ provider: agentProviderSchema, status: z.literal('unavailable'), reason: z.string(), checked_at_ms: checkedAtSchema }),
  z.object({ provider: agentProviderSchema, status: z.literal('not_configured'), reason: z.string(), checked_at_ms: checkedAtSchema }),
]);

export type ProviderAvailability = z.infer<typeof providerAvailabilitySchema>;

export const agentProvidersSchema = z.array(providerAvailabilitySchema);

/** `GET /api/agent-providers`; `recheck` runs every check again instead of reading the server's 30 s cache. */
export function agentProvidersOperation(recheck: boolean): ApiOperation<ProviderAvailability[]> {
  return {
    method: 'GET',
    path: recheck ? '/api/agent-providers?refresh=true' : '/api/agent-providers',
    responseSchema: agentProvidersSchema,
  };
}

/** `provider`'s entry, or `null` when the answer is not in yet or does not name it. */
export function availabilityOf(
  answers: readonly ProviderAvailability[] | undefined,
  provider: AgentProvider,
): ProviderAvailability | null {
  return answers?.find((answer) => answer.provider === provider) ?? null;
}

File: crates/calm-server/src/routes/agent_providers.rs
//! `GET /api/agent-providers` — whether each Planner provider can run right now (#1817).

use axum::extract::{Query, State};
use axum::{Json, Router, routing::get};
use serde::Deserialize;
use utoipa::IntoParams;

use crate::agent_providers::{Freshness, ProviderAvailability};
use crate::error::{ErrorBody, Result};
use crate::state::{AppState, CodexShellState, RouteState};

#[derive(Debug, Deserialize, IntoParams)]
pub struct AgentProvidersQuery {
    /// `true` runs every check again instead of answering from the last one (at most 30 s old).
    #[serde(default)]
    pub refresh: bool,
}

pub fn router() -> Router<AppState> {
    Router::new().route("/api/agent-providers", get(list_agent_providers))
}

#[utoipa::path(
    get,
    path = "/api/agent-providers",
    tag = "agent_providers",
    params(AgentProvidersQuery),
    responses(
        (status = 200, description = "One entry per Planner provider, Codex then Claude. `ready` passed every check; `unavailable` names the first failed check and its fix in `reason`; `not_configured` is a provider this server was not started with (Claude without `--claude-planner-config`). Codex: the shared app-server runs and `account/read` says it is logged in. Claude: the pinned binary answers `--version` with the pinned version and `auth status --json` says `loggedIn`. Answers are cached per provider for 30 s; a failed check is an answer, never an error.", body = [ProviderAvailability]),
        (status = 500, description = "Internal error", body = ErrorBody),
    ),
)]
pub(crate) async fn list_agent_providers(
    State(s): State<RouteState>,
    State(codex): State<CodexShellState>,
    Query(q): Query<AgentProvidersQuery>,
) -> Result<Json<Vec<ProviderAvailability>>> {
    let freshness = if q.refresh {
        Freshness::Recheck
    } else {
        Freshness::Cached
    };
    let checked = s
        .provider_availability
        .all(freshness, &s.claude_planner, &codex.shared_codex_appserver)
        .await;
    Ok(Json(
        checked
            .into_iter()
            .map(ProviderAvailability::from)
            .collect(),
    ))
}
